### PR TITLE
[metal] Complete Device API

### PR DESCRIPTION
**Pull request by k-ye**
Related issue = https://github.com/taichi-dev/taichi/issues/2736

### CODE REVIEW

File: taichi/backends/metal/device.cpp [hunk 1]
```diff
@@ -34,7 +34,8 @@ class ResourceBinderImpl : public ResourceBinder {
                  uint32_t binding,
                  DevicePtr ptr,
                  size_t size) override {
-    TI_NOT_IMPLEMENTED;
+    TI_ERROR(
```
Comment (bobcao3) [inline taichi/backends/metal/device.cpp:37]: Is this a TODO or a not supported feature in Metal?

Comment (k-ye) [reply taichi/backends/metal/device.cpp:37]: TODO

Comment (k-ye) [reply taichi/backends/metal/device.cpp:37]: More specifically, Metal has this API where you can set `offset`, but not `size`... https://developer.apple.com/documentation/metal/mtlcomputecommandencoder/1443126-setbuffer?language=objc

Comment (bobcao3) [reply taichi/backends/metal/device.cpp:37]: This is actually fine, we can ignore size.

Comment (bobcao3) [reply taichi/backends/metal/device.cpp:37]: In vulkan when offset+size go over the buffer size it's undefined behavior anyways

Comment (k-ye) [reply taichi/backends/metal/device.cpp]: Fixed

**Review by bobcao3 — APPROVED** [review]
LGTM

**Pull request merged**